feat(crypto): implement multilinear sumcheck data structures

ashjeong opened the pull request:
Creates needed data structures for multilinear sumcheck such as the verifying key and a class dedicated for sumcheck evaluation. Additionally tests the data structures and their corresponding methods.

Commit:
test(zk): add unit tests for `ConstraintSystem`

I've referred to lookups in simple_lookup_circuit.h and [dev.rs](https://github.com/kroma-network/halo2/blob/7d0a36990452c8e7ebd600de258420781a9b7917/halo2_proofs/src/dev.rs#L1595-L1895).

Commit:
perf(c/zk): read only verifying key when replaying verifier

Commit:
fix(zk): read the commitments to the advice columns in the correct order

Unlike other columns, the advice columns have their own phases. The phase
iterates from 0, and the advice columns are committed when the current phase
is equal to the phases of the advice columns. Therefore, commitments
are created in the order of phase. However, commitments should be written
to the proof in the order of column. For example, if there are commitments
{C₀, C₁, C₂} corresponding to columns 0, 1, 2 and phases {0, 1, 0},
previously, the proof was read in the order of C₀, C₂, C₁,
even though the correct order should have been C₀, C₁, C₂.
In this commit, the order to read the columns has been corrected.

chokobole committed:
test(zk): add parallelization benchmark

chokobole committed:
perf(zk): move `permutations` when calling `BuildProvingKey()`

chokobole committed:
perf(zk): deallocate fixed and instance columns vec before building circuit polynomial

chokobole committed:
perf(c/math): copy points in parallel

chokobole committed:
refac(zk): rename `CompressLookupStep` to `PermuteLooksStep`

chokobole committed:
refac(zk): rename `CoeffsToExtendedPart` to `CoeffsToExtendedParts`

chokobole committed:
test(zk): remove redundant unittest

chokobole committed:
refac(zk): remove the trivial return type bool

chokobole committed:
refac(zk): rename variable name for consistency with others

chokobole committed:
perf(zk): deallocate large vector immediately after it is used

chokobole committed:
perf(zk): optimize `CompressExpressions()`

1. This commit uses `OPENMP_PARALLEL_FOR` instead of `base::Parallelize()`
   which allocates intermediate vectors.
2. This commit moves scalar multiplication and additions to an existing
   parallel loop which saves costs of creating and joining threads.
3. This commit uses a single buffer instead of two buffers.

chokobole committed:
perf(zk): reserve `repeated_input_rows`

chokobole committed:
chore(zk): add missing comments

chokobole committed:
refac(zk): add missing changes from #253

chokobole committed:
style(zk): rename `lookup_argument_unittests` to `lookup_unittests`

chokobole committed:
refac(zk): use usable rows instead of blinding factors

chokobole committed:
refac(zk): reorder method arguments to conform with any other methods in `GrandProductArgument`

chokobole committed:
perf(zk): use a single buffer to create z(X) in `GrandProductArgument`

chokobole committed:
chore(zk): add a missing period

chokobole committed:
perf(zk): optimize `ProvingKey::DoLoad()`

1. Use `OPENMP_PARALLEL_FOR` instead of `base::Parallelize()` which
   allocates intermediate memory space.
2. Iterate parallel for loop using `usable_rows` instead of
   the entire domain size.

chokobole committed:
perf(zk): parallelize `UnpermutedTable::Construct()`

chokobole committed:
refac: use `base::Map` to create vector

chokobole committed:
feat(zk): set the proof cursor to "done" after finishing reading the proof

chokobole committed:
chore(zk): remove TODO comment since `CommitLagrange` is already run in parallel

chokobole committed:
refac: remove unused header `base/containers/adapters.h`

chokobole committed:
perf: remove intermediate memory allocation from `base::ParallelizeXXX`

chokobole committed:
fix(math): fix `PointXYZZ::BatchNormalize()`

We should've checked whether z² is one, not z⁻³ is one.

chokobole committed:
perf(math): enable `BatchMapScalarField()` and `BatchNormalize()` to run in parallel

ashjeong committed:
feat: add multilinear sumcheck data structures

Referenced [arkworks' sumcheck code](https://github.com/arkworks-rs/sumcheck/blob/master/src/ml_sumcheck/data_structures.rs)

chokobole committed:
chore(zk): remove the obsolete comment

chokobole committed:
test(c/math): use the number of coefficients instead of degree

chokobole committed:
test(zk): remove redundant test from `PermutationArgumentTest`

chokobole committed:
test(crypto): initialize the prime field within the test where it is used

chokobole committed:
test: add a common base class `FiniteFieldTest`

chokobole committed:
test(c/math): add a common base class `c::math::bn254::G1Test`

chokobole committed:
test(zk): use smaller `GF7` instead of `bn254::Fr`

chokobole committed:
test(zk): avoid using `halo2::ProverTest`

chokobole committed:
test(zk): use minimum required header

chokobole committed:
test(zk): rename `CompressExpressionTestSetting` to `CompressExpressionTest`

chokobole committed:
refac(zk): remove unused method

chokobole committed:
feat(zk): add mutable getter to `LookupPair`

chokobole committed:
chore(zk): add TODO comment

chokobole commented on tachyon/crypto/sumcheck/multilinear/BUILD.bazel:
```diff
@@ -0,0 +1,14 @@
+package(default_visibility = ["//visibility:public"])
+
+tachyon_cc_library(
+    name = "multilinear_sumcheck",
```
Are you going to add more stuffs here? Otherwise, please name it `data_structures`.

```suggestion
    name = "data_structures",
```

ashjeong replied:
The actual multilinear sumcheck will be implemented like [so](https://github.com/arkworks-rs/sumcheck/blob/master/src/ml_sumcheck/mod.rs) later, but you're right. I'll change this to "data_structures" and create a different tachyon_cc_library label later for the actual multilinear_sumcheck.

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,173 @@
+
+namespace crypto::sumcheck {
+
+// This data structure  will be used as the verifier key.
```
```suggestion
// This data structure will be used as the verifier key.
```

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,173 @@
+
+namespace base {
+
+using PolynomialInfo = crypto::sumcheck::PolynomialInfo;
```
Please do not create type alias in header. This might cause unexpected symbol errors.

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,173 @@
+template <>
+class Copyable<PolynomialInfo> {
+ public:
+  static bool WriteTo(const PolynomialInfo& bigint, Buffer* buffer) {
```
```suggestion
  static bool WriteTo(const PolynomialInfo& poly_info, Buffer* buffer) {
```

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,173 @@
+
+// NOTE(ashjeong): univariate_evaluations.h used as reference
+template <typename H, typename F, typename Evaluations>
+H AbslHashValue(H h, const math::MultilinearExtension<Evaluations>* evals) {
```
Add testing for hash! Please check `ref_unittest.cc`

ashjeong replied:
Looking back, I don't think I need this hash function, since it's already implemented for MultilinearDenseEvaluations/MultilinearExtensions. I've gotten rid of it and the code works fine too!

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,173 @@
+
+// NOTE(ashjeong): univariate_evaluations.h used as reference
+template <typename H, typename F, typename Evaluations>
+H AbslHashValue(H h, const math::MultilinearExtension<Evaluations>* evals) {
```
```suggestion
H AbslHashValue(H h, const math::MultilinearExtension<Evaluations>& evals) {
```

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,173 @@
+
+  // Adds a Term to the list
+  void AddTerm(
+      F& coefficient,
+      std::vector<std::shared_ptr<math::MultilinearExtension<Evaluations>>>&
+          evaluations) {
```
```suggestion
      const F& coefficient,
      const std::vector<std::shared_ptr<math::MultilinearExtension<Evaluations>>>&
          evaluations) {
```

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,173 @@
+  }
+
+  // Evaluates all terms together on a an evaluation point for each variable:
+  // $$\sum_{i=0}^{n}coefficient_i\cdot\prod_{j=0}^{m_i}evaluation_{ij}$$
```
Please update this comment using unicode not latex.

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,173 @@
+                     term.indexes.begin(), term.indexes.end(), F::One(),
+                     [this, &point](F& product, size_t index) {
+                       return product *=
+                              this->flattened_ml_evaluations_[index]->Evaluate(
```
```suggestion
                              flattened_ml_evaluations_[index]->Evaluate(
```

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,173 @@
+        indexed_evaluations.push_back(curr_index);
+      }
+    }
+    terms_.push_back(Term{coefficient, indexed_evaluations});
```
```suggestion
    terms_.push_back({coefficient, indexed_evaluations});
```

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,173 @@
+      CHECK_EQ(evaluation->Degree(), num_variables_);
+      const math::MultilinearExtension<Evaluations>* evaluation_ptr =
+          evaluation.get();
+      auto it = raw_pointers_lookup_table_.find(evaluation_ptr);
+      if (it != raw_pointers_lookup_table_.end()) {
```
Please use `base::Contains`

```suggestion
      if (base::Contains(raw_pointers_lookup_table_, evaluation_ptr))
```

Commit:
perf(math): change `kParallelBatchInverseDivisorThreshold` to 2

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,173 @@
+  // max number of evaluations for each term
+  size_t max_evaluations_ = 0;
+  // number of variables of the polynomial
+  // Note(ashjeong): this is equivalent to `kMaxDegree` of multilinear
+  // extensions
+  size_t num_variables_ = 0;
```
If so, why do we have this member?

ashjeong replied:
It's needed to check line 75: `CHECK_EQ(evaluation->Degree(), num_variables_);` (note evaluation->Degree() gives us kMaxDegree for the given evaluation) and ensure that the number of variables in each individual evaluation given is valid. I put that note, since when I look at the MultilinearExtension or DenseMultilinearEvaluation files I was confused because the equivalent value of the intuitively labelled variable `num_variable` was instead non-intuitively named `kMaxDegree`.

chokobole replied:
Q1. Why does `evaluation->Degree()` return `kMaxDegree`? Shouldn't it return the degree, not the maximum degree?
Q2. Can you explain why `kMaxDegree` is named in a non-intuitive manner?

Commit:
perf(zk): take the return value as a reference instead of copying it

Commit:
perf(zk): copy only one polynomial when constructing h poly

Commit:
refac(zk): remove unused header

Commit:
perf(zk): avoid copying

Commit:
perf(zk): avoid memory reallocation

Commit:
perf(zk): remove redundant parallel loop

Commit:
perf: use `OPENMP_PARALEL_NESTED_FOR` for nested loop

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,167 @@
+  std::vector<std::shared_ptr<math::MultilinearExtension<Evaluations>>>
+  flattened_ml_evaluations() {
+    return flattened_ml_evaluations_;
+  }
```
```suggestion
  size_t max_evaluations() const { return max_evaluations_; }
  size_t num_variables() const { return num_variables_; }
  const std::vector<Term>& terms() const { return terms_; }
  const std::vector<std::shared_ptr<math::MultilinearExtension<Evaluations>>>&
  flattened_ml_evaluations() const {
    return flattened_ml_evaluations_;
  }
```

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,167 @@
+ public:
+  using F = typename math::MultilinearExtension<Evaluations>::Field;
+
+  struct Term {
+    F coefficient;
+    std::vector<size_t> indexes;
+  };
```
Please move this out of the `ListOfProductsOfPolynomials` because this class needs only `F`. Please see `UnivariateTerm` in `univariate_sparse_coefficients.h`

ashjeong replied:
I'm not sure this should be moved out, since the elements in the vector `indexes` rely on member `flattened_ml_extensions` within the `ListOfProductsOfPolynomials` class.

chokobole replied:
Got it!

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,167 @@
+#include "tachyon/math/polynomials/multivariate/multilinear_extension.h"
+
+namespace tachyon {
+
```
Please remove the blank here.

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,167 @@
+class ListOfProductsOfPolynomials;
+
+template <typename Evaluations>
+class ListOfProductsOfPolynomials<math::MultilinearExtension<Evaluations>> {
```
Shouldn't it be in math namespace?

chokobole replied:
Additionally, I think `SumOfProductOfPoly` or `LinearCombination` is more appropriate name for this class

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,167 @@
+namespace crypto::sumcheck {
+
+// This data structure will be used as the verifier key.
+struct TACHYON_EXPORT PolynomialInfo {
```
Shouldn't it be math namespace?

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures_unittest.cc:
```diff
@@ -0,0 +1,119 @@
+    math::MultilinearDenseEvaluations<math::GF7, kMaxDegree>>;
+using Evals = math::MultilinearDenseEvaluations<math::GF7, kMaxDegree>;
+using ProductList = ListOfProductsOfPolynomials<Poly>;
+using Point = math::GF7;
```
```suggestion
using Point = Poly::Point;
```

ashjeong replied:
Yep that makes sense. I also added a separate `using F = math:GF7`.

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,167 @@
+               const std::vector<
+                   std::shared_ptr<math::MultilinearExtension<Evaluations>>>&
+                   evaluations) {
+    CHECK(!evaluations.empty());
```
Why do you kill the program? What about just skipping like below?

```suggestion
if (evaluations.empty()) return;
```

ashjeong replied:
I followed the Rust code, but simply returning like you did makes much more sense!

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,167 @@
+    indexed_evaluations.reserve(num_evaluations);
+    for (std::shared_ptr<math::MultilinearExtension<Evaluations>> evaluation :
+         evaluations) {
+      CHECK_EQ(evaluation->Degree(), num_variables_);
```
Why do we check this? Because `evaluation->Degree()` obviously should be lower than `kMaxDegree`?

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,167 @@
+  }
+
+  // Evaluates all terms together on a an evaluation point for each variable:
+  // ∑ᵢ₌₀ⁿ(coefficientᵢ⋅∏ⱼ₌₀ᵐevaluationᵢⱼ)
```
I believe this looks better!

```suggestion
// ∑ᵢ₌₀..ₙ(coefficientᵢ⋅∏ⱼ₌₀..ₘevaluationᵢⱼ)
```

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,167 @@
+  // Evaluates all terms together on a an evaluation point for each variable:
+  // ∑ᵢ₌₀ⁿ(coefficientᵢ⋅∏ⱼ₌₀ᵐevaluationᵢⱼ)
+  // where n = total terms, m = number of evaluations per term
+  F Evaluate(std::vector<F>& point) const {
```
```suggestion
  template <typename Container>
  F Evaluate(const Container& point) const {
```

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,167 @@
+        [this, &point](F& sum, const Term& term) {
+          return sum +=
+                 term.coefficient *
+                 std::accumulate(
+                     term.indexes.begin(), term.indexes.end(), F::One(),
+                     [this, &point](F& product, size_t index) {
+                       return product *=
+                              flattened_ml_evaluations_[index]->Evaluate(point);
+                     });
```
Could you change it to the method of `Term`?

chokobole replied:
And I think we can parallelize! Please refer to `MultivariateSparseCoefficients::Evaluate()`.

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,167 @@
+  // where n = total terms, m = number of evaluations per term
+  F Evaluate(std::vector<F>& point) const {
+    CHECK_EQ(point.size(), num_variables_);
+    return std::accumulate(
+        terms_.begin(), terms_.end(), F::Zero(),
+        [this, &point](F& sum, const Term& term) {
+          return sum +=
+                 term.coefficient *
```
Certainly, it depends on the expected number of terms. If there are many, please parallelize it.

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,167 @@
+  }
+
+  // Extract the max number of evaluations and number of variables
+  PolynomialInfo Info() const { return {max_evaluations_, num_variables_}; }
```
If this method returns required getters, why do we need this method?

ashjeong replied:
I'm conflicted about this too... the getters are used in the inner setup for the prover side of the sumcheck protocol, but `PolynomialInfo` is used for the verifier side of the sumcheck protocol 🤔. Possibly I should separate them to correlate specially as so.

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,167 @@
+  }
+
+  // Adds a Term to the list
+  void AddTerm(const F& coefficient,
+               const std::vector<
+                   std::shared_ptr<math::MultilinearExtension<Evaluations>>>&
+                   evaluations) {
```
Do you think we don't need a method `AddTerm(const Term&)`?

ashjeong replied:
That is a fair point-I'll definitely add that.

ashjeong replied:
Actually, I don't think that's a good idea to add, since `Term`s have a member of `indexes` that rely on the member `flattened_ml_extensions` in `LinearCombinations`, so it doesn't make sense.

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,167 @@
+    max_evaluations_ = std::max(max_evaluations_, num_evaluations);
+    std::vector<size_t> indexed_evaluations;
+    indexed_evaluations.reserve(num_evaluations);
+    for (std::shared_ptr<math::MultilinearExtension<Evaluations>> evaluation :
+         evaluations) {
```
```suggestion
    for (const std::shared_ptr<math::MultilinearExtension<Evaluations>>& evaluation :
         evaluations) {
```

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,167 @@
+#include "absl/container/node_hash_map.h"
+
+#include "tachyon/base/buffer/copyable.h"
+#include "tachyon/base/containers/contains.h"
```
Please remove this! I shared you with a better idea!

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,167 @@
+      CHECK_EQ(evaluation->Degree(), num_variables_);
+      const math::MultilinearExtension<Evaluations>* evaluation_ptr =
+          evaluation.get();
+      if (base::Contains(raw_pointers_lookup_table_, evaluation_ptr)) {
+        indexed_evaluations.push_back(
+            raw_pointers_lookup_table_[evaluation_ptr]);
+      } else {
+        size_t curr_index = flattened_ml_evaluations_.size();
+        flattened_ml_evaluations_.push_back(evaluation);
+        raw_pointers_lookup_table_[evaluation_ptr] = curr_index;
+        indexed_evaluations.push_back(curr_index);
+      }
```
```suggestion
      auto it = raw_pointers_lookup_table_.try_emplace(
          evaluation_ptr, flattened_ml_evaluations_.size());
      // if inserted
      if (it.second) {
        flattened_ml_evaluations_.push_back(evaluation);
      }
      indexed_evaluations.push_back((*it.first).second);
```

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,167 @@
+        indexed_evaluations.push_back(curr_index);
+      }
+    }
+    terms_.push_back({coefficient, indexed_evaluations});
```
```suggestion
    terms_.push_back({coefficient, std::move(indexed_evaluations)});
```

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,167 @@
+    CHECK(!evaluations.empty());
+    size_t num_evaluations = evaluations.size();
+    max_evaluations_ = std::max(max_evaluations_, num_evaluations);
+    std::vector<size_t> indexed_evaluations;
```
Could you rename it to `indexes` to conform with the member of `Term`?

chokobole commented on tachyon/crypto/sumcheck/multilinear/data_structures.h:
```diff
@@ -0,0 +1,167 @@
+  // Adds a Term to the list
+  void AddTerm(const F& coefficient,
+               const std::vector<
+                   std::shared_ptr<math::MultilinearExtension<Evaluations>>>&
```
Throughout the codes, we rarely uses `std::shared_ptr` so I am kind of curious how `std::shared_ptr` gets used after sumcheck!

ashjeong replied:
Looking at all the other code involved in the arkworks sumcheck repository, I don't really think the shared_ptrs are used again 🤔. On the prover side of the protocol, `flattened_ml_extensions_` is deep copied, but is instead used as a vector of `MultilinearExtensions` and not of shared_ptrs. I tried showing a better general view of what goes on in the protocol more concretely here, though I still don't understand the Prover/Verifier side fully yet. Hopefully the diagram helps a bit to understand the structure of the protocol.
![image](https://github.com/ashjeong/tachyon/assets/84874402/d15c8bb3-e4ae-4498-98c4-2fe0f4331fdc)

Commit:
Merge pull request #289 from kroma-network/fix/read-commitments-from-proof-in-the-correct-order

fix(zk): read commitments from proof in the correct order

Commit:
feat(zk): terminate the program if `WriteToProof()` is failed

Related: #239

Commit:
feat(crypto): add `ToString()` and `ToHexString()` to `PolynomialOpening`

Commit:
perf(zk): apply lazy IFFT and commit

1. Halo2 holds the polynomials in both coefficient and evaluation form.
   Considering that the polynomials in coefficient form are needed after y,
   we can delay IFFT until phase y. This can help reduce memory pressure.
2. The compressed pairs are needed to construct grand product polys.
   The h poly and evals are needed to construct combined h poly.
   After they are used, this commit immediately releases their memory.
3. Since the conversion from bucket to affine point requires
   an inverse operation, `BatchNormalize()` is used to reduce
   the number of required inverse operations. This commit tries to
   use `BatchNormalize()` on every commitment after phase theta.

Commit:
test(crypto): add missing changes from #211

Commit:
feat(crypto): implement the equality operator on polynomial opening structures

Commit:
test(crypto): reinforce the unittest on top of #274

Commit:
feat(crypto): add `GroupBySinglePoint()` to `PolynomialOpeningGrouper`

See
[construct_intermediate_sets](https://github.com/kroma-network/halo2/blob/7d0a36990452c8e7ebd600de258420781a9b7917/halo2_proofs/src/poly/kzg/multiopen/gwc.rs#L36-L61).

Commit:
feat: change `CreateOpeningProof` to mutable method

This is needed if the opening scheme uses `BatchCommit` inside.

dongchangYoo committed:
perf(zk): merge separated parallelized parts in `CircuitPolynomialBuilder`

Commit:
perf(math): remove redundant for loop

Commit:
refac(crypto): reorder the left and right inputs to pairing

This is to conform with the order of the paring inputs in halo2.

See
[DualMSM::check()](https://github.com/kroma-network/halo2/blob/7d0a36990452c8e7ebd600de258420781a9b7917/halo2_proofs/src/poly/kzg/msm.rs#L174-L194).

Commit:
chore(math): update comments

Commit:
feat(math): use `std::size()` to get the size of the container

Commit:
feat(crypto): implement `GWC` opening scheme

See
[gwc](https://github.com/kroma-network/halo2/tree/7d0a36990452c8e7ebd600de258420781a9b7917/halo2_proofs/src/poly/kzg/multiopen/gwc).

Commit:
test(zk): separate the concrete class `BN254SHPlonkProverTest` from `ProverTest`

Commit:
chore(zk): update comment

1. change URL comment to be one line.
2. fix typo.

Commit:
refac(zk): make circuit include only `zk/plonk/constarint_system/circuit.h`

Commit:
build(zk): add a missing dep

Commit:
test(zk): make circuit independent of shplonk

Commit:
refac(zk): group FRIEND_TEST by the test type and sort them

Commit:
fix(zk): fix compile error when `PCS::kQueryInstance` is enabled

Commit:
refac(zk): add missing `const`

ashjeong committed:
Merge pull request #290 from kroma-network/perf/optimize-halo2-prover

perf: optimize halo2 prover

Commit:
test(zk): add `GWC` prove & verify test to `ShuffleCircuitTest`

Commit:
Merge pull request #295 from kroma-network/perf/remove-redundant-for-loop

perf(math): remove redundant for loop

Commit:
perf(zk): move shared values out of for loop

Commit:
test(zk): prevent the creation of a global variable for `std::vector<T>`

See
https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables.

Commit:
Merge pull request #291 from kroma-network/test/improve-test

test: improve test

Commit:
refac(zk): move halo2 coupled lookup codes to `zk/lookup/halo2`

dongchangYoo committed:
perf(zk): move input of `FFT/IFFT` into the function where possible

Commit:
test(zk): remove redundant unittest

Commit:
refac(c/zk): use switch statement

Commit:
refac: apply c++17 fold expressions to `base::EstimateSize()`

Commit:
test: use `Uint8VectorBuffer::Grow()` instead of injecting the buffer directly

Commit:
feat(math): add `Copyable` to eigen matrix

Commit:
perf(math): set the default strategy to `kParallelTerm`

Commit:
fix(math): set zero if the inputs of msm are empty

Commit:
fix(math): fix bug computing size of `results`

This commit also refactors the codes to follow the style in
`base/parallelize.h`.

Commit:
feat(crypto): implement the equality operator on `PoseidonConfig(Entry)`

Commit:
feat(base): add `Copyable` to enum type

Commit:
feat(crypto): add `Copyable` to `PoseidonSponge`

Commit:
feat(crypto): implement the equality operator on `PoseidonSponge`

Commit:
feat: add `GetDigestLen()` and `GetStateLen()` to `Blake2bWriter`

Commit:
refac(c/zk): move `TACHYON_HALO2_BLAKE_TRANSCRIPT` to `constants.h`

This constant has nothing to do with the curve type!

Commit:
chore(zk): update comments

Commit:
refac: move `TranscriptType` to `zk/plonk/halo2`

Commit:
feat: support poseidon transcript

Commit:
feat(zk): remove unused methods from rust binding side

Related: #249

Commit:
refac(zk): use `WriteMany()` and `ReadMany()`

dongchangYoo committed:
Merge pull request #294 from kroma-network/perf/optimize-halo2-prover-2

Commit:
feat: support sha256 transcript

ashjeong committed:
Merge pull request #299 from kroma-network/perf/set-the-default-strategy-to-parallel-term

perf: set the default strategy to parallel term

dongchangYoo committed:
Merge pull request #297 from kroma-network/perf/merge-parallelized-parts

perf(zk): merge parallelized parts

Commit:
Merge pull request #296 from kroma-network/feat/implement-gwc-opening-scheme

feat: implement gwc opening scheme

dongchangYoo committed:
feat(math): add FFT/IFFT that takes input through move

dongchangYoo committed:
perf(zk): merge separated parallelized parts related to`GrandProductArgument`

Commit:
refac(crypto): removed unused variables

Commit:
Merge pull request #298 from kroma-network/perf/move-input-to-IFFT/FFT

perf(zk): move input to ifft/fft

Commit:
perf(math): optimize `DoBatchMapScalarFieldToPoint`

Commit:
perf(math): optimize `EvaluatePartialLagrangeCoefficients`

Commit:
Merge pull request #302 from kroma-network/perf/merge-parallelized-parts-2

perf(zk): merge parallelized parts related to grand product

Commit:
fix: correct conversion from wide bytes to prime field

We interpreted `from_bytes_wide` wrong and made `FromAnySizedBigInt` and
used it somewhere. This commit has replaced it with a correct version.

dongchangYoo committed:
Merge pull request #300 from kroma-network/feat/add-transcript-c-apis-for-kroma-prover

feat: add transcript c apis for kroma prover

Commit:
test(math): reinforce test for parallel IFFT

Commit:
Merge pull request #287 from kroma-network/test/add-cs-unittests

test(zk): add unit tests for `ConstraintSystem`

ashjeong committed:
test: test multilinear sumcheck data structures

ashjeong committed:
Merge pull request #309 from kroma-network/test/reinforce-ifft-test

ashjeong committed:
Merge pull request #303 from kroma-network/perf/optimize-unsafe-setup

perf: optimize `KZG::UnsafeSetup`
